### PR TITLE
ReportsConfig Componentization

### PR DESCRIPTION
**Pull request by Guillar1**
Planning on removing the mock service worker in the tests in another PR.

### CODE REVIEW

File: packages/itwin/reports-config-widget/src/widget/components/ActionPanel.tsx [hunk 1]
```diff
@@ -35,16 +34,15 @@ const ActionPanel = ({
       >
         {actionLabel}
       </Button>
-      <Button
-        styleType="default"
-        type="button"
+      {onCancel && <Button
+        styleType='default'
+        type='button'
+        id='cancel'
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/ActionPanel.tsx:40]: Why do you need the id here? Such a generic id is almost a guarantee to conflict with some other id. Ids should be unique throughout the whole application.

File: packages/itwin/reports-config-widget/src/widget/components/ActionPanel.tsx [hunk 2]
```diff
@@ -35,16 +34,15 @@ const ActionPanel = ({
...
         onClick={onCancel}
         disabled={isCancelDisabled || isLoading}
       >
-        {ReportsConfigWidget.localization.getLocalizedString(
-          "ReportsConfigWidget:Cancel"
-        )}
-      </Button>
+        Cancel
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/ActionPanel.tsx:44]: Why was localization removed?

File: packages/itwin/reports-config-widget/src/widget/context/ReportsApiConfigContext.tsx [hunk 3]
```diff
@@ -6,14 +6,20 @@ import type { AccessToken } from "@itwin/core-bentley";
 import * as React from "react";
 import { createContext } from "react";
 
+export type GetAccessTokenFn = () => Promise<AccessToken>;
+
 export interface ReportsApiConfig {
-  getAccessToken: () => Promise<AccessToken>;
+  getAccessToken: GetAccessTokenFn;
+  iTwinId: string;
+  iModelId: string;
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/context/ReportsApiConfigContext.tsx:14]: Why is iModelId required for ReportsAPI?

File: packages/itwin/reports-config-widget/src/widget/components/SearchBar.tsx [hunk 4]
```diff
@@ -20,29 +20,17 @@ export const SearchBar = ({
...
-      }}
-      onAnimationEnd={() => {
-        if (searchBarClosing) {
-          setSearchBarClosing(false);
-          setSearchBarOpened(false);
-          setSearchValue("");
-        }
-      }}
-    >
+    <div className="rcw-search-button">
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/SearchBar.tsx:25]: Looks like there is a visual change (removed animation) and functional change (removed search value reset). Why is this not in a separate PR? Why such change even needs to be made?

File: packages/itwin/reports-config-widget/src/widget/components/Reports.tsx [hunk 5]
```diff
@@ -125,140 +106,123 @@ export const Reports = () => {
...
+              />
+            ))}
+          </div>
+        )}
+      </Surface>
+      <DeleteModal
+        entityName={showDeleteModal?.displayName}
+        onClose={() => setShowDeleteModal(undefined)}
+        onDelete={async () => {
+          const reportsClientApi = new ReportsClient(
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/Reports.tsx:215]: I see that ReportsClient is getting created in multiple places. Maybe it could be handled in a context somewhere? That way you will not need to pass around the baseUrl.

File: packages/itwin/reports-config-widget/src/widget/components/AddMappingsModal.tsx [hunk 6]
```diff
@@ -43,35 +44,36 @@ const fetchMappings = async (
...
 export const AddMappingsModal = ({
   reportId,
   existingMappings,
   show,
-  returnFn,
+  onClose,
 }: AddMappingsModalProps) => {
   const [isLoading, setIsLoading] = useState<boolean>(false);
-  const [selectedMappings, setSelectedMappings] = useState<Mapping[]>([]);
+  const selectedMappings = useRef<Mapping[]>([]);
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/AddMappingsModal.tsx:61]: Why the change from state to ref?

File: packages/itwin/reports-config-widget/src/widget/components/ActionPanel.tsx (hunk 1 above)
Comment (Guillar1) [reply packages/itwin/reports-config-widget/src/widget/components/ActionPanel.tsx]: id removed. Was not needed.

File: packages/itwin/reports-config-widget/src/widget/components/ActionPanel.tsx (hunk 2 above)
Comment (Guillar1) [reply packages/itwin/reports-config-widget/src/widget/components/ActionPanel.tsx]: Brought back.

File: packages/itwin/reports-config-widget/src/widget/context/ReportsApiConfigContext.tsx (hunk 3 above)
Comment (Guillar1) [reply packages/itwin/reports-config-widget/src/widget/context/ReportsApiConfigContext.tsx]: I've renamed the context to `ReportsConfigContext` for broader generalization. Additionally, I removed `iModelId` since it was only utilized in one instance: during the addition of Mappings to the report. Previously, the `iModelId` was used to default-select the list of iModels from which to choose mappings.

File: packages/itwin/reports-config-widget/src/widget/components/SearchBar.tsx (hunk 4 above)
Comment (Guillar1) [reply packages/itwin/reports-config-widget/src/widget/components/SearchBar.tsx]: I will leave this to another PR. I was having issues with the animation.

File: packages/itwin/reports-config-widget/src/widget/components/Reports.tsx (hunk 5 above)
Comment (Guillar1) [reply packages/itwin/reports-config-widget/src/widget/components/Reports.tsx]: I move all the client creation to the context. baseUrl is still needed in one location. Which is to generate the `odataFeedUrl`.

File: packages/itwin/reports-config-widget/src/widget/components/AddMappingsModal.tsx (hunk 6 above)
Comment (Guillar1) [reply packages/itwin/reports-config-widget/src/widget/components/AddMappingsModal.tsx:59]: For now, I am no longer using the useState to render the selected mappings. Just to store them. It will save some unnecessary re-renders.

File: packages/itwin/reports-config-widget/src/test/ReportAction.test.tsx [hunk 7]
```diff
@@ -5,87 +5,29 @@
...
-  selectionManagerMock
-    .setup((x) => x.scopes)
-    .returns(() => selectionScopesManagerMock.object);
-
-  Presentation.setSelectionManager(selectionManagerMock.object);
-  await TestUtils.initializeUiFramework(connectionMock.object);
-  await ReportsConfigWidget.initialize();
+  const localization = new EmptyLocalization();
+  await ReportsConfigWidget.initialize(localization);
   server.listen();
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/test/ReportAction.test.tsx:27]: What is the server used for? It should not be required if all clients are mocked.

File: packages/itwin/reports-config-widget/src/test/test-utils.tsx [hunk 8]
```diff
@@ -4,35 +4,27 @@
...
   const Wrapper = ({ children }: WrapperProps) => (
-    <ReportsApiConfigContext.Provider
-      value={{
-        getAccessToken: mockAccessToken,
-        baseUrl: REPORTS_CONFIG_BASE_URL,
-      }}
+    <ReportsConfigContext
+      getAccessToken={mockAccessToken}
+      iTwinId={contextProps?.iTwinId ?? mockITwinId}
+      bulkExtractor={contextProps?.bulkExtractor}
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/test/test-utils.tsx:23]: I don't think you need to specify `bulkExtractor` explicitly here. The line below should take care of it.

File: packages/itwin/reports-config-widget/src/test/ReportAction.test.tsx (hunk 7 above)
Comment (Guillar1) [reply packages/itwin/reports-config-widget/src/test/ReportAction.test.tsx:27]: Only one test required it. I quickly removed it.

File: packages/itwin/reports-config-widget/src/test/test-utils.tsx (hunk 8 above)
Comment (Guillar1) [reply packages/itwin/reports-config-widget/src/test/test-utils.tsx:23]: Yup, don't need it. Forgot to remove it. There used to be a default mock here.

File: packages/itwin/reports-config-widget/src/widget/components/AddMappingsModal.tsx (hunk 6 above)
Comment (MindaugasButkus) [reply packages/itwin/reports-config-widget/src/widget/components/AddMappingsModal.tsx:59]: So how does the UI know what mappings are selected? I think we might encounter hard-to-debug bugs related to this.

File: packages/itwin/reports-config-widget/src/widget/components/ReportsRouter.tsx [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+export const ReportsRouter = ({
+  currentRoute,
+  navigateTo,
+  goBack,
+}: {
+  currentRoute: Route;
+  navigateTo: (toRoute: (prev: Route | undefined) => Route) => void;
+  goBack: () => void;
+}) => {
+  goBack;
```
Comment (MindaugasButkus) [inline packages/itwin/reports-config-widget/src/widget/components/ReportsRouter.tsx:23]: What is this for?

File: packages/itwin/reports-config-widget/src/test/ReportAction.test.tsx (hunk 7 above)
Comment (MindaugasButkus) [reply packages/itwin/reports-config-widget/src/test/ReportAction.test.tsx:27]: What do you mean by "removed it"? It's still there.

File: packages/itwin/reports-config-widget/src/widget/components/ReportsRouter.tsx (hunk 9 above)
Comment (Guillar1) [reply packages/itwin/reports-config-widget/src/widget/components/ReportsRouter.tsx]: Nothing just a noop. Was a debug attempt. I removed it earlier. Pushed my changes.

File: packages/itwin/reports-config-widget/src/test/ReportAction.test.tsx (hunk 7 above)
Comment (Guillar1) [reply packages/itwin/reports-config-widget/src/test/ReportAction.test.tsx]: Gone.

File: packages/itwin/reports-config-widget/src/widget/components/AddMappingsModal.tsx (hunk 6 above)
Comment (Guillar1) [reply packages/itwin/reports-config-widget/src/widget/components/AddMappingsModal.tsx:59]: The table itself keeps track of this.

**Review by MindaugasButkus — APPROVED** [review]
